File: Samples/CompositionVisual/cpp/VisualProperties.cpp
#include "pch.h"
#include "VisualProperties.h"

#include "windowsnumerics.h"

using namespace Windows::ApplicationModel::Core;
using namespace Windows::UI::Core;
using namespace Windows::Foundation;
using namespace Windows::UI::Composition;
using namespace Windows::Foundation::Numerics;
using namespace Windows::UI;

//------------------------------------------------------------------------------
//
// main
//
//------------------------------------------------------------------------------


// The main function is only used to initialize our IFrameworkView class.
[Platform::MTAThread]
int main(Platform::Array<Platform::String^>^)
{
	auto visualPropertiesFactory = ref new VisualPropertiesFactory();
	CoreApplication::Run(visualPropertiesFactory);
	return 0;
}


VisualProperties::VisualProperties()
{
}

//------------------------------------------------------------------------------
//
// VisualProperties.Initialize
//
// This method is called during startup to associate the IFrameworkView with the
// CoreApplicationView.
//
//------------------------------------------------------------------------------

void VisualProperties::Initialize(CoreApplicationView^ view)
{
	_view = view;
}

//------------------------------------------------------------------------------
//
// VisualProperties.SetWindow
//
// This method is called when the CoreApplication has created a new CoreWindow,
// allowing the application to configure the window and start producing content
// to display.
//
//------------------------------------------------------------------------------

void VisualProperties::SetWindow(Windows::UI::Core::CoreWindow^ window)
{
	_window = window;
	InitNewComposition();

	_window->PointerPressed += ref new TypedEventHandler<CoreWindow^, PointerEventArgs^>(this, &VisualProperties::OnPointerPressed);
	_window->PointerMoved += ref new TypedEventHandler<CoreWindow^, PointerEventArgs^>(this, &VisualProperties::OnPointerMoved);
	_window->PointerReleased += ref new TypedEventHandler<CoreWindow^, PointerEventArgs^>(this, &VisualProperties::OnPointerReleased);
}

//------------------------------------------------------------------------------
//
// VisualProperties.OnPointerPressed
//
// This method is called when the user touches the screen, taps it with a stylus
// or clicks the mouse.
//
//------------------------------------------------------------------------------

void VisualProperties::OnPointerPressed(Windows::UI::Core::CoreWindow^ window, Windows::UI::Core::PointerEventArgs^ args)
{
	auto position = args->CurrentPoint->Position;

	//
	// Walk our list of visuals to determine who, if anybody, was selected
	//
	for (Visual^ child : _root->Children)
	{
		//
		// Did we hit this child?
		//
		float3 offset = child->Offset;
		float2 size = child->Size;

		if ((position.X >= offset.x) &&
			(position.X < offset.x + size.x) &&
			(position.Y >= offset.y) &&
			(position.Y < offset.y + size.y))
		{
			//
			// This child was hit. Since the children are stored back to front,
			// the last one hit is the front-most one so it wins
			//
			_currentVisual = dynamic_cast<ContainerVisual^>(child);
			_offsetBias = new float2((float)(offset.x - position.X),
				(float)(offset.y - position.Y));
		}
	}

	//
	// If a visual was hit, bring it to the front of the Z order
	//
	if (_currentVisual != nullptr)
	{
		ContainerVisual^ parent = dynamic_cast<ContainerVisual^>(_currentVisual->Parent);
		parent->Children->Remove(_currentVisual);
		parent->Children->InsertAtTop(_currentVisual);
	}
}

//------------------------------------------------------------------------------
//
// VisualProperties.OnPointerMoved
//
// This method is called when the user moves their finger, stylus or mouse with
// a button pressed over the screen.
//
//------------------------------------------------------------------------------

void VisualProperties::OnPointerMoved(Windows::UI::Core::CoreWindow^ window, Windows::UI::Core::PointerEventArgs^ args)
{
	//
	// If a visual is selected, drag it with the pointer position and
	// make it opaque while we drag it
	//
	if (_currentVisual != nullptr)
	{
		//
		// Set up the properties of the visual the first time it is
		// dragged. This will last for the duration of the drag
		//
		if (!_dragging)
		{
			_currentVisual->Opacity = 1.0f;

			//
			// Transform the first child of the current visual so that
			// the image is rotated
			//
            for (auto child : _currentVisual->Children)
			{
				child->RotationAngleInDegrees = 45.0f;
				child->CenterPoint = float3(_currentVisual->Size.x / 2, _currentVisual->Size.y / 2, 0);
                break;
			}

			//
			// Clip the visual to its original layout rect by using an inset
			// clip with a one-pixel margin all around
			//
			auto clip = _compositor->CreateInsetClip();
			clip->LeftInset = 1.0f;
			clip->RightInset = 1.0f;
			clip->TopInset = 1.0f;
			clip->BottomInset = 1.0f;
			_currentVisual->Clip = clip;

			_dragging = true;
		}

		Point position = args->CurrentPoint->Position;
		_currentVisual->Offset = float3((float)(position.X + _offsetBias->x),
			(float)(position.Y + _offsetBias->y),
			0.0f);
	}
}

//------------------------------------------------------------------------------
//
// VisualProperties.OnPointerReleased
//
// This method is called when the user lifts their finger or stylus from the
// screen, or lifts the mouse button.
//
//------------------------------------------------------------------------------

void VisualProperties::OnPointerReleased(Windows::UI::Core::CoreWindow^ window, Windows::UI::Core::PointerEventArgs^ args)
{
	//
	// If a visual was selected, make it transparent again when it is
	// released and restore the transform and clip
	//
	if (_currentVisual != nullptr)
	{
		if (_dragging)
		{
			//
			// Remove the transform from the first child
			//
            for (auto child : _currentVisual->Children)
			{
				child->RotationAngleInDegrees = 0;
				child->CenterPoint = float3(0, 0, 0);
                break;
			}

			_currentVisual->Opacity = 0.8f;
			_currentVisual->Clip = nullptr;
			_dragging = false;
		}

		_currentVisual = nullptr;
	}
}

//------------------------------------------------------------------------------
//
// VisualProperties.InitNewComposition
//
// This method is called by SetWindow(), where we initialize Composition after
// the CoreWindow has been created.
//
//------------------------------------------------------------------------------

void VisualProperties::InitNewComposition()
{
	//
	// Set up Windows.UI.Composition Compositor, root ContainerVisual, and associate with
	// the CoreWindow.
	//

	_compositor = ref new Compositor();

	_root = _compositor->CreateContainerVisual();

	_compositionTarget = _compositor->CreateTargetForCurrentView();
	_compositionTarget->Root = _root;

	//
	// Create a few visuals for our window
	//
	for (int index = 0; index < 20; index++)
	{
		_root->Children->InsertAtTop(CreateChildElement());
	}
}

//------------------------------------------------------------------------------
//
// VisualProperties.CreateChildElement
//
// Creates a small sub-tree to represent a visible element in our application.
//
//------------------------------------------------------------------------------

Visual^ VisualProperties::CreateChildElement()
{
	//
	// Each element consists of three visuals, which produce the appearance
	// of a framed rectangle
	//
	auto element = _compositor->CreateContainerVisual();
	element->Size = float2(100.0f, 100.0f);

	//
	// Position this visual randomly within our window
	//
	element->Offset = float3(((float)rand() / RAND_MAX) * 400, ((float)rand() / RAND_MAX) * 400, 0.0f);

	//
	// The outer rectangle is always white
	//
	auto visual = _compositor->CreateSpriteVisual();
	element->Children->InsertAtTop(visual);

	visual->Brush = _compositor->CreateColorBrush(ColorHelper::FromArgb(0xFF, 0xFF, 0xFF, 0xFF));
	visual->Size = float2(100.0f, 100.0f);

	//
	// The inner rectangle is inset from the outer by three pixels all around
	//
	auto child = _compositor->CreateSpriteVisual();
	visual->Children->InsertAtTop(child);
	child->Offset = float3(3.0f, 3.0f, 0.0f);
	child->Size = float2(94.0f, 94.0f);

	//
	// Pick a random color for every rectangle
	//
	byte red = (byte)(0xFF * (0.2f + (((float)rand() / RAND_MAX) / 0.8f)));
	byte green = (byte)(0xFF * (0.2f + (((float)rand() / RAND_MAX) / 0.8f)));
	byte blue = (byte)(0xFF * (0.2f + (((float)rand() / RAND_MAX) / 0.8f)));
	child->Brush = _compositor->CreateColorBrush(ColorHelper::FromArgb(0xFF, red, green, blue));

	//
	// Make the subtree root visual partially transparent. This will cause each visual in the subtree
	// to render partially transparent, since a visual's opacity is multiplied with its parent's
	// opacity
	//
	element->Opacity = 0.8f;

	return element;
}

//------------------------------------------------------------------------------
//
// VisualProperties.Load
//
// This method is called when a specific page is being loaded in the
// application.  It is not used for this application.
//
//------------------------------------------------------------------------------

void VisualProperties::Load(Platform::String ^ unused)
{
}

//------------------------------------------------------------------------------
//
// VisualProperties.Run
//
// This method is called by CoreApplication.Run() to actually run the
// dispatcher's message pump.
//
//------------------------------------------------------------------------------

void VisualProperties::Run()
{
	_window->Activate();
	_window->Dispatcher->ProcessEvents(CoreProcessEventsOption::ProcessUntilQuit);
}

//------------------------------------------------------------------------------
//
// VisualProperties.Uninitialize
//
// This method is called during shutdown to disconnect the CoreApplicationView,
// and CoreWindow from the IFrameworkView.
//
//------------------------------------------------------------------------------

void VisualProperties::Uninitialize()
{
	_window = nullptr;
	_view = nullptr;
}

File: Samples/CompositionVisual/vb/VisualProperties.vb
'*********************************************************
'
' Copyright (c) Microsoft. All rights reserved.
' This code is licensed under the MIT License (MIT).
' THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, 
' INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
' FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
' IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, 
' DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, 
' TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH 
' THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
'
'*********************************************************
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Numerics
Imports System.Text
Imports System.Threading.Tasks
Imports Windows.ApplicationModel.Core
Imports Windows.Foundation
Imports Windows.UI
Imports Windows.UI.Composition
Imports Windows.UI.Core

Namespace Global.CompositionVisual

    Class VisualProperties
        Implements IFrameworkView

        Sub Initialize(view As CoreApplicationView) Implements IFrameworkView.Initialize
            _view = view
            _random = New Random()
        End Sub

        Sub SetWindow(window As CoreWindow) Implements IFrameworkView.SetWindow
            _window = window
            InitNewComposition()
            AddHandler _window.PointerPressed, AddressOf OnPointerPressed
            AddHandler _window.PointerMoved, AddressOf OnPointerMoved
            AddHandler _window.PointerReleased, AddressOf OnPointerReleased
        End Sub

        Sub OnPointerPressed(window As CoreWindow, args As PointerEventArgs)
            Dim position As Point = args.CurrentPoint.Position
            For Each child In _root.Children
                '
                ' Did we hit this child?
                '
                Dim offset As Vector3 = child.Offset
                Dim size As Vector2 = child.Size
                If(position.X >= offset.X) AndAlso (position.X < offset.X + size.X) AndAlso (position.Y >= offset.Y) AndAlso (position.Y < offset.Y + size.Y) Then
                    _currentVisual = TryCast(child, ContainerVisual)
                    _offsetBias = New Vector2(CType((offset.X - position.X), Single), CType((offset.Y - position.Y), Single))
                End If
            Next

            If _currentVisual IsNot Nothing Then
                Dim parent As ContainerVisual = TryCast(_currentVisual.Parent, ContainerVisual)
                parent.Children.Remove(_currentVisual)
                parent.Children.InsertAtTop(_currentVisual)
            End If
        End Sub

        Sub OnPointerMoved(window As CoreWindow, args As PointerEventArgs)
            If _currentVisual IsNot Nothing Then
                If Not _dragging Then
                    _currentVisual.Opacity = 1.0f
                    For Each child In _currentVisual.Children
                        child.RotationAngleInDegrees = 45.0F
                        child.CenterPoint = New Vector3(_currentVisual.Size.X / 2, _currentVisual.Size.Y / 2, 0)
                        Exit For
                    Next

                    '
                    ' Clip the visual to its original layout rect by using an inset
                    ' clip with a one-pixel margin all around
                    '
                    Dim clip = _compositor.CreateInsetClip()
                    clip.LeftInset = 1.0f
                    clip.RightInset = 1.0f
                    clip.TopInset = 1.0f
                    clip.BottomInset = 1.0f
                    _currentVisual.Clip = clip
                    _dragging = True
                End If

                Dim position As Point = args.CurrentPoint.Position
                _currentVisual.Offset = New Vector3(CType((position.X + _offsetBias.X), Single), CType((position.Y + _offsetBias.Y), Single), 0.0f)
            End If
        End Sub

        Sub OnPointerReleased(window As CoreWindow, args As PointerEventArgs)
            If _currentVisual IsNot Nothing Then
                If _dragging Then
                    For Each child In _currentVisual.Children
                        child.RotationAngle = 0.0f
                        child.CenterPoint = New Vector3(0.0f, 0.0f, 0.0f)
                        Exit For
                    Next

                    _currentVisual.Opacity = 0.8f
                    _currentVisual.Clip = Nothing
                    _dragging = False
                End If

                _currentVisual = Nothing
            End If
        End Sub

        Sub Load(unused As String) Implements IFrameworkView.Load
        End Sub

        Sub Run() Implements IFrameworkView.Run
            _window.Activate()
            _window.Dispatcher.ProcessEvents(CoreProcessEventsOption.ProcessUntilQuit)
        End Sub

        Sub Uninitialize() Implements IFrameworkView.Uninitialize
            _window = Nothing
            _view = Nothing
        End Sub

        Sub InitNewComposition()
            _compositor = New Compositor()
            _root = _compositor.CreateContainerVisual()
            _compositionTarget = _compositor.CreateTargetForCurrentView()
            _compositionTarget.Root = _root
            For index = 0 To 20 - 1
                _root.Children.InsertAtTop(CreateChildElement())
            Next
        End Sub

        Function CreateChildElement() As Visual
            '
            ' Each element consists of three visuals, which produce the appearance
            ' of a framed rectangle
            '
            Dim element = _compositor.CreateContainerVisual()
            element.Size = New Vector2(100.0f, 100.0f)
            element.Offset = New Vector3(CType((_random.NextDouble() * 400), Single), CType((_random.NextDouble() * 400), Single), 0.0f)
            '
            ' The outer rectangle is always white
            '
            Dim visual = _compositor.CreateSpriteVisual()
            element.Children.InsertAtTop(visual)
            visual.Brush = _compositor.CreateColorBrush(Color.FromArgb(&HFF, &HFF, &HFF, &HFF))
            visual.Size = New Vector2(100.0f, 100.0f)
            '
            ' The inner rectangle is inset from the outer by three pixels all around
            '
            Dim child = _compositor.CreateSpriteVisual()
            visual.Children.InsertAtTop(child)
            child.Offset = New Vector3(3.0f, 3.0f, 0.0f)
            child.Size = New Vector2(94.0f, 94.0f)
            '
            ' Pick a random color for every rectangle
            '
            Dim red As Byte = CType((&HFF * (0.2f + (_random.NextDouble() / 0.8f))), Byte)
            Dim green As Byte = CType((&HFF * (0.2f + (_random.NextDouble() / 0.8f))), Byte)
            Dim blue As Byte = CType((&HFF * (0.2f + (_random.NextDouble() / 0.8f))), Byte)
            child.Brush = _compositor.CreateColorBrush(Color.FromArgb(&HFF, red, green, blue))
            element.Opacity = 0.8f
            Return element
        End Function

        ' CoreWindow / CoreApplicationView
        Private _window As CoreWindow

        Private _view As CoreApplicationView

        ' Windows.UI.Composition
        Private _compositor As Compositor

        Private _compositionTarget As CompositionTarget

        Private _root As ContainerVisual

        Private _currentVisual As ContainerVisual

        Private _offsetBias As Vector2

        Private _dragging As Boolean

        ' Helpers
        Private _random As Random
    End Class

    Public NotInheritable Class VisualPropertiesFactory
        Implements IFrameworkViewSource

        Function CreateView() As IFrameworkView Implements IFrameworkViewSource.CreateView
            Return New VisualProperties()
        End Function

        '------------------------------------------------------------------------------
        '
        ' main
        '
        '------------------------------------------------------------------------------
        Shared Function Main(args As String()) As Integer
            CoreApplication.Run(New VisualPropertiesFactory())
            Return 0
        End Function
    End Class
End Namespace
